crates/hir_ty/src/infer.rs

```diff
@@ -78,6 +78,19 @@ pub(crate) fn infer_query(db: &dyn HirDatabase, def: DefWithBodyId) -> Arc<Infer
     Arc::new(ctx.resolve_all())
 }
 
+pub(crate) fn can_unify(db: &dyn HirDatabase, def: DefWithBodyId, ty1: Ty, ty2: Ty) -> bool {
+    let resolver = def.resolver(db.upcast());
+    let mut ctx = InferenceContext::new(db, def, resolver);
+
+    let ty1 = ctx.canonicalizer().canonicalize_ty(ty1).value;
+    let ty2 = ctx.canonicalizer().canonicalize_ty(ty2).value;
```

Comment:
I think that for IDE purposes, you should be able to just assume the two `Ty`s to already be canonicalized since they shouldn't contain any type variables. That should allow you to do this without needing the DB or def.

Reply:
Removed this function (see prev comment).